File: frontend/lib/features/home/presentation/greetings.dart
import '../../../l10n/app_localizations.dart';
import '../../../shared/widgets/typewriter_text.dart';

/// Salutul rotativ din bara de sus a paginii principale, construit din ora și
/// data de pe telefon.
///
/// Ordinea nu e întâmplătoare: mai întâi ocaziile speciale (ziua ta, sărbători),
/// pentru că sunt cele care surprind; apoi mesajele potrivite cu ora; la final
/// contextul de zi a săptămânii. `TypewriterText` le rulează în buclă, deci
/// primul din listă e și primul văzut.
///
/// Funcția e pură - primește `now`, nu îl citește singură - ca să poată fi
/// testată pe orice moment din an fără să ne atingem de ceasul sistemului.
List<TypewriterPhrase> buildGreetings({
  required AppLocalizations l10n,
  required DateTime now,
  String? name,
  int? birthdayMonth,
  int? birthdayDay,
}) {
  final displayName = (name != null && name.trim().isNotEmpty)
      ? name.trim()
      : l10n.greetReaderFallback;

  const occasionHold = Duration(seconds: 8);
  const normalHold = Duration(seconds: 3);

  final phrases = <TypewriterPhrase>[
    for (final text in _occasions(l10n, now, birthdayMonth, birthdayDay))
      TypewriterPhrase(text, occasionHold),
    for (final text in _forHour(l10n, now.hour, displayName))
      TypewriterPhrase(text, normalHold),
    for (final text in _forWeekday(l10n, now)) TypewriterPhrase(text, normalHold),
  ];

  // Prima frază stă mai mult: e cea pe care o vede oricine deschide aplicația
  // și pleacă în câteva secunde.
  return [
    TypewriterPhrase(phrases.first.text, const Duration(seconds: 10)),
    ...phrases.skip(1),
  ];
}

/// Sărbători și aniversări. Pot fi mai multe într-o zi (ex. ziua ta pe 1
/// Decembrie), deci întoarcem o listă, nu o singură valoare.
List<String> _occasions(
  AppLocalizations l10n,
  DateTime now,
  int? birthdayMonth,
  int? birthdayDay,
) {
  final occasions = <String>[];

  if (birthdayMonth == now.month && birthdayDay == now.day) {
    occasions.add(l10n.greetBirthday);
  }
  // 1 Decembrie - Ziua Națională.
  if (now.month == 12 && now.day == 1) {
    occasions.add(l10n.greetNationalDay);
  }
  // Crăciunul ține trei zile: ajunul plus prima și a doua zi.
  if (now.month == 12 && now.day >= 24 && now.day <= 26) {
    occasions.add(l10n.greetChristmas);
  }
  // Anul Nou: 31 decembrie și primele două zile din ianuarie.
  if ((now.month == 12 && now.day == 31) ||
      (now.month == 1 && now.day <= 2)) {
    occasions.add(l10n.greetNewYear);
  }
  // 23 aprilie - Ziua Internațională a Cărții.
  if (now.month == 4 && now.day == 23) {
    occasions.add(l10n.greetBookDay);
  }

  final easter = orthodoxEaster(now.year);
  final easterMonday = easter.add(const Duration(days: 1));
  if (_sameDay(now, easter) || _sameDay(now, easterMonday)) {
    occasions.add(l10n.greetEaster);
  }

  return occasions;
}

/// Cele cinci intervale de zi cerute în Milestone 14.
List<String> _forHour(AppLocalizations l10n, int hour, String name) {
  if (hour >= 5 && hour < 10) {
    return [
      l10n.greetMorningSun,
      l10n.greetMorningNamed(name),
      l10n.greetMorningCoffeeBook,
      l10n.greetMorningCoffee,
      l10n.greetMorningStartStory,
      l10n.greetMorningAdventure,
      l10n.greetMorningSleptWell,
      l10n.greetMorningPerfectBook,
      l10n.greetMorningNicer,
    ];
  }
  if (hour >= 10 && hour < 17) {
    return [
      l10n.greetDayNamed(name),
      l10n.greetDayDiscover,
      l10n.greetDayAdventure,
      l10n.greetDayLibrary,
      l10n.greetDayCorporateCoffee,
      l10n.greetDayWhatsNext,
      l10n.greetDaySwappedToday,
      l10n.greetDayNewReader,
      l10n.greetDayFindNext,
    ];
  }
  if (hour >= 17 && hour < 21) {
    return [
      l10n.greetEveningHello,
      l10n.greetEveningHowWasDay,
      l10n.greetEveningPerfectNow,
      l10n.greetEveningFewPages,
      l10n.greetEveningRelax,
      l10n.greetEveningQuiet,
      l10n.greetEveningWhatTonight,
      l10n.greetEveningBeforeBed,
    ];
  }
  if (hour >= 21) {
    return [
      l10n.greetNightGoodNight,
      l10n.greetNightSandman,
      l10n.greetNightCloseBook,
      l10n.greetNightSleepWell(name),
      l10n.greetNightOneMoreChapter,
      l10n.greetNightSeeYouTomorrow,
      l10n.greetNightNiceDay,
      l10n.greetNightQuiet,
    ];
  }
  // 00:00 - 05:00
  return [
    l10n.greetLateAwake,
    l10n.greetLateNightOwl,
    l10n.greetLateMidnightReads,
    l10n.greetLateOneChapter,
    l10n.greetLateNeverSleeps,
    l10n.greetLateCompany,
    l10n.greetLateLastChapter,
    l10n.greetLateForgotSleep,
  ];
}

List<String> _forWeekday(AppLocalizations l10n, DateTime now) {
  switch (now.weekday) {
    case DateTime.saturday:
    case DateTime.sunday:
      return [l10n.greetWeekend];
    case DateTime.monday:
      return [l10n.greetMonday];
    case DateTime.friday:
      // Doar seara - vineri la 9 dimineața mesajul n-are sens.
      return now.hour >= 17 ? [l10n.greetFridayEvening] : const [];
    default:
      return const [];
  }
}

bool _sameDay(DateTime a, DateTime b) =>
    a.year == b.year && a.month == b.month && a.day == b.day;

/// Data Paștelui ortodox (cel sărbătorit în România), ca dată gregoriană.
///
/// Algoritmul lui Meeus pentru calendarul iulian, plus decalajul iulian →
/// gregorian. Decalajul e 13 zile pentru intervalul 1900-2099, ceea ce acoperă
/// orice an în care aplicația asta va rula; în afara lui rezultatul ar fi
/// greșit cu o zi sau mai mult.
DateTime orthodoxEaster(int year) {
  final a = year % 4;
  final b = year % 7;
  final c = year % 19;
  final d = (19 * c + 15) % 30;
  final e = (2 * a + 4 * b - d + 34) % 7;
  final month = (d + e + 114) ~/ 31; // 3 = martie, 4 = aprilie
  final day = ((d + e + 114) % 31) + 1;
  return DateTime(year, month, day).add(const Duration(days: 13));
}

File: web/src/features/home/greetings.ts
import type { TypewriterPhrase } from '@/components/ui/TypewriterText';

type Translate = (key: string, params?: Record<string, unknown>) => string;

const OCCASION_HOLD = 8000;
const NORMAL_HOLD = 3000;
const MOTTO_HOLD = 5000;

/** Prima frază stă mai mult: e cea pe care o citește oricine deschide aplicația. */
const FIRST_HOLD = 10000;

const MORNING = [
  'greetMorningSun',
  'greetMorningNamed',
  'greetMorningCoffeeBook',
  'greetMorningCoffee',
  'greetMorningStartStory',
  'greetMorningAdventure',
  'greetMorningSleptWell',
  'greetMorningPerfectBook',
  'greetMorningNicer',
];

const DAY = [
  'greetDayNamed',
  'greetDayDiscover',
  'greetDayAdventure',
  'greetDayLibrary',
  'greetDayCorporateCoffee',
  'greetDayWhatsNext',
  'greetDaySwappedToday',
  'greetDayNewReader',
  'greetDayFindNext',
];

const EVENING = [
  'greetEveningHello',
  'greetEveningHowWasDay',
  'greetEveningPerfectNow',
  'greetEveningFewPages',
  'greetEveningRelax',
  'greetEveningQuiet',
  'greetEveningWhatTonight',
  'greetEveningBeforeBed',
];

const NIGHT = [
  'greetNightGoodNight',
  'greetNightSandman',
  'greetNightCloseBook',
  'greetNightSleepWell',
  'greetNightOneMoreChapter',
  'greetNightSeeYouTomorrow',
  'greetNightNiceDay',
  'greetNightQuiet',
];

const MOTTOS = [
  'greetMottoStandingTree',
  'greetMottoWhyBuyNew',
  'greetMottoMoreSustainable',
  'greetMottoEuropeanMovement',
  'greetMottoCirculating',
];

/**
 * Frazele de salut din antetul paginii principale. Port al
 * home/presentation/greetings.dart.
 *
 * `now` e ora LOCALĂ a dispozitivului, nu a serverului: salutul trebuie să
 * urmeze fusul userului, altfel „bună dimineața" ar apărea la miezul nopții
 * pentru cine e în alt fus.
 */
export function buildGreetings({
  t,
  now,
  name,
  birthdayMonth,
  birthdayDay,
}: {
  t: Translate;
  now: Date;
  name?: string | null;
  birthdayMonth?: number | null;
  birthdayDay?: number | null;
}): TypewriterPhrase[] {
  const displayName = name?.trim() ? name.trim() : t('greetReaderFallback');

  const phrases: TypewriterPhrase[] = [
    ...occasions(t, now, birthdayMonth, birthdayDay).map((text) => ({
      text,
      holdMs: OCCASION_HOLD,
    })),
    ...forHour(t, now.getHours(), displayName).map((text) => ({
      text,
      holdMs: NORMAL_HOLD,
    })),
    ...forWeekday(t, now).map((text) => ({ text, holdMs: NORMAL_HOLD })),
    ...MOTTOS.map((key) => ({ text: t(key), holdMs: MOTTO_HOLD })),
  ];

  if (phrases.length === 0) return [];
  return [{ text: phrases[0].text, holdMs: FIRST_HOLD }, ...phrases.slice(1)];
}

function occasions(
  t: Translate,
  now: Date,
  birthdayMonth?: number | null,
  birthdayDay?: number | null,
): string[] {
  const month = now.getMonth() + 1;
  const day = now.getDate();
  const result: string[] = [];

  if (birthdayMonth === month && birthdayDay === day) result.push(t('greetBirthday'));
  if (month === 12 && day === 1) result.push(t('greetNationalDay'));
  if (month === 12 && day >= 24 && day <= 26) result.push(t('greetChristmas'));
  if ((month === 12 && day === 31) || (month === 1 && day <= 2)) {
    result.push(t('greetNewYear'));
  }
  if (month === 4 && day === 23) result.push(t('greetBookDay'));

  const easter = orthodoxEaster(now.getFullYear());
  const easterMonday = new Date(easter);
  easterMonday.setDate(easterMonday.getDate() + 1);
  if (sameDay(now, easter) || sameDay(now, easterMonday)) result.push(t('greetEaster'));

  return result;
}

function forHour(t: Translate, hour: number, name: string): string[] {
  // Cheile „Named" primesc numele; restul îl ignoră. Îl trimitem la toate -
  // ICU ignoră un parametru nefolosit, iar alternativa ar fi o listă de
  // excepții care se desincronizează la prima cheie adăugată.
  const translate = (key: string) => t(key, { name });

  if (hour >= 5 && hour < 10) return MORNING.map(translate);
  if (hour >= 10 && hour < 17) return DAY.map(translate);
  if (hour >= 17 && hour < 21) return EVENING.map(translate);
  return NIGHT.map(translate);
}

function forWeekday(t: Translate, now: Date): string[] {
  const weekday = now.getDay();
  if (weekday === 0 || weekday === 6) return [t('greetWeekend')];
  if (weekday === 1) return [t('greetMonday')];
  if (weekday === 5 && now.getHours() >= 17) return [t('greetFridayEvening')];
  return [];
}

function sameDay(a: Date, b: Date): boolean {
  return (
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate()
  );
}

/**
 * Paștele ortodox (Meeus, calendar iulian convertit la gregorian). E singura
 * sărbătoare din listă care se mută de la an la an, deci nu poate fi o dată
 * fixă ca celelalte.
 */
function orthodoxEaster(year: number): Date {
  const a = year % 4;
  const b = year % 7;
  const c = year % 19;
  const d = (19 * c + 15) % 30;
  const e = (2 * a + 4 * b - d + 34) % 7;
  const month = Math.floor((d + e + 114) / 31);
  const day = ((d + e + 114) % 31) + 1;

  // Rezultatul e în calendarul iulian; decalajul față de cel gregorian e de
  // 13 zile pentru secolele XX-XXI.
  const julian = new Date(year, month - 1, day);
  julian.setDate(julian.getDate() + 13);
  return julian;
}
